README
Hermes is a very small, very simple fileserver. 
Forked from https://github.com/joshualockhart/hermes

To host a file, do

$ ./hermes my_file

then to retrieve it on another host, do something like

$ nc <server host> 9999 > my_file

Hermes defaults to TCP/9999, but the port can be specified with -p.
If you want a (limited) amount of verbosity, use -v.

Joshua Lockhart wrote this for fun. 
If you like it then please feel free to email him at joshualockhart ( AT ) gmail ( DOT ) com. 

If you like a portion of the code and want to copy and paste it to use in your project, please email or fork him with a copy of said project. 
He won't be upset, just interested in what you're working on. :)


ChangeLog
LiDE "Linux Disk Extractor"

As of version 0.0.1.1 it has the same functionality as running a string of commands:
-dd to capture a file (more like cat though)
-pv "pipe viewer" to view the progress of the file transfer
-netcat to send the file over the network
All packaged into a small executable less than 15Kb in size.

The following command could accomplish the same results:
cat File_to_Transfer | pv | nc -l -p 9999


Summary ChangeLog

Original source code: 
https://github.com/joshualockhart/Hermes
by Josh Lockhart, joshualockhart@gmail.com, 30 Oct 2010

Fixed Original Source:
-Fixed hang, on end of file transfer.
-Submited Pull request to https://github.com/joshualockhart/Hermes

v0.0.1.1
-Added ChangeLog
-Renamed project to LiME
-Renamed files.
-Fixed version.
-Patched Makefile

v0.0.1.2
-Removed option for verbose, always verbose.
-Changed default port.
-Removed new line for progress status (verbose).


Referenced Sources

https://github.com/joshualockhart/hermes
Thanks to Joshua Lockhart for getting my project off the ground. 


LiDE.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/types.h>
#include "LiDE.h"

void print_help(void){
	printf("LiDE - v0.0.1.1 \n \
			-h: Prints this help.\n \
			-v: Be verbose.\n \
			-p: Specify a port to listen on. Default 9999.\n");
	exit(0);
}

int main(int argc, char *argv[])
{

	be_verbose = 0;
	FILE *filename_fd;
	char *filename_to_serve;
	int listen_port = htons(9999);

	if (argc < 2)
		print_help();

	char c;
	while ((c = getopt(argc, argv, "hvp:")) != -1) {
		switch(c){

			case 'h':
				print_help();
				break;
			case 'v':
				printf("Being verbose.\n");
				be_verbose = 1;
				break;
			case 'p':
				if ((optarg != NULL) && (optarg[0] != '\0')){

					listen_port = htons(atoi(optarg));

					if (listen_port < 1024 && getuid() > 0){
						printf("Invalid port number. Binding to a port lower than 1024 requires superuser priveleges. You, unfortunately, are not so priveleged. Sorry!\n");
						exit(0);
					}
					else {
						listen_port = htons(atoi(optarg));
						break;
					}
				}
				else break;
			default:
				print_help();
				break;
		}
	}

	// Get the filename, which is the argument without an option.
	if ((argc - optind) == 1)
	{
		if (strlen(argv[argc-1]) > 255)
		{
			printf("Filename too long.\n");
			exit(0);
		}

		if((filename_to_serve = malloc(strlen(argv[argc-1]) + 1)) == NULL)
		{
			perror("Malloc");
			exit(1);
		}
		strncpy(filename_to_serve, argv[argc-1], strlen(argv[argc-1]));
	}


	if ((filename_fd = fopen(filename_to_serve, "r")) != NULL){
		listen_for_inbound_requests(listen_port, filename_fd);
		printf("%d\n", listen_port);
	}

	else
		perror("Error opening file");

	free(filename_to_serve);


	return 0;
}


LiME.c
#include <stdio.h>        // C library to perform Input/Output operations
#include <unistd.h>       //
#include <string.h>       // C Strings
#include <stdlib.h>       // C Standard General Utilities Library
#include <arpa/inet.h>    //
#include <sys/socket.h>   //
#include <sys/types.h>    //
#include "LiDE.h"         //

void print_help(void){
	printf("\x1B[33mLiDE (\033[1mLi\033[22mnux \033[1mD\033[22misk \033[1mE\033[22mxtractor) - v0.0.1.5\033[0m\n\n\
LiDE is a very small fileserver.\n\
To host a file, do\n\n\
$ \x1B[32m./lide <filename>\033[0m              Will use default port 31337\n\
    or\n\
$ \x1B[32m./lide -p <port> <filename>\033[0m    To specify a port to send on.\n\n\
$ \x1B[32m./lide [-h]\033[0m                    Prints this help.\n\n\
Then to retrieve it on another host, do something like\n\n\
$ \x1B[32mnc <LiDE Host IP> 31337 > retrieved_filename\033[0m\n");
	exit(0);
}

int main(int argc, char *argv[])
{

	be_verbose = 1;
	FILE *filename_fd;
	char *filename_to_serve;
	int listen_port = htons(31337);

	if (argc < 2)
		print_help();

	char c;
	while ((c = getopt(argc, argv, "hp:")) != -1) {
		switch(c){

			case 'h':
				print_help();
				break;
			case 'p':
				if ((optarg != NULL) && (optarg[0] != '\0')){

					listen_port = htons(atoi(optarg));

					if (listen_port < 1024 && getuid() > 0){
						printf("Invalid port number. Binding to a port lower than 1024 requires root priveleges.\n");
						exit(0);
					}
					else {
						listen_port = htons(atoi(optarg));
						break;
					}
				}
				else break;
			default:
				print_help();
				break;
		}
	}

	// Get the filename, which is the argument without an option.
	if ((argc - optind) == 1)
	{
		if (strlen(argv[argc-1]) > 255)
		{
			printf("Filename too long.\n");
			exit(0);
		}

		if((filename_to_serve = malloc(strlen(argv[argc-1]) + 1)) == NULL)
		{
			perror("Malloc");
			exit(1);
		}
		strncpy(filename_to_serve, argv[argc-1], strlen(argv[argc-1]));
	}


	if ((filename_fd = fopen(filename_to_serve, "r")) != NULL){
		listen_for_inbound_requests(listen_port, filename_fd);
		printf("%d\n", listen_port);
	}

	else
		perror("Error opening file");

	free(filename_to_serve);


	return 0;
}



nwio.c
#include <stdio.h>        //
#include <unistd.h>       //
#include <string.h>       //
#include <stdlib.h>       //
#include <arpa/inet.h>    //
#include <sys/socket.h>   //
#include "LiDE.h"         //

/*
 * Send the file to the currently connected client, return number of bytes sent or 0 on error. At the minute
 * if there is an error with a send call it just dies and returns -1, this could obviously be improved to 
 * make it a little "smarter".
 */

int send_file(FILE *fd, int connfd)
{
	int filesize,
	    number_complete_chunks,
	    final_chunk_size,
	    total_bytes_sent = 0,
	    bytes_sent = 0,
	    i;

	fseek(fd, 0L, SEEK_END);
	filesize = ftell(fd);
	rewind(fd);

	number_complete_chunks = filesize / CHUNK_SIZE;
	final_chunk_size = filesize % CHUNK_SIZE;

	char *chunk;

	if((chunk = (char *) malloc(CHUNK_SIZE)) == NULL)
	{
		perror("Malloc");
		exit(1);
	}


	for (i = 0; i < number_complete_chunks; i++)
	{
		fread(chunk, CHUNK_SIZE, 1, fd);

		if ((bytes_sent = send(connfd, chunk, CHUNK_SIZE, 0)) < 0)
			return -1;

		total_bytes_sent += bytes_sent;
		if(be_verbose)
			printf("\x1B[31m%d\x1B[32m / %d\040bytes\r\e[?25l\033[0m", total_bytes_sent, filesize); // the "\r" refreshes current line and "\e[?25l" removes cursor
	}

	// send the last incomplete chunk at the end of the file.
	fread(chunk, final_chunk_size, 1, fd);

	if ((bytes_sent = send(connfd, chunk, final_chunk_size, 0)) < 0)
		return -1;

	total_bytes_sent += bytes_sent;
	if(be_verbose)
		printf("\x1B[32m%d / %d\040bytes\033[0m\n", total_bytes_sent, filesize);

	free(chunk);
	return total_bytes_sent;
}

/*
 * Wait for someone to ask for our file.
 */

int listen_for_inbound_requests(int port, FILE *filename_fd)
{
	int listenfd, bytes_sent;
	int connfd;
	struct sockaddr_in serv, cli;	
	socklen_t len;

	char *ip_buff; // Somewhere handy to put IP addresses.
	
	if((ip_buff = (char *) malloc(INET_ADDRSTRLEN)) == NULL)
	{
		perror("Malloc");
		exit(1);
	}

	if ((listenfd = socket(AF_INET, SOCK_STREAM, 0)) == -1){
		perror("Error creating socket");
	}

	int tr = 1;
	if (setsockopt(listenfd,SOL_SOCKET,SO_REUSEADDR,&tr,sizeof(int)) == -1) { 
		perror("setsockopt");
		exit(1);
	}

	bzero(&serv, sizeof(serv));
	serv.sin_family = AF_INET;
	serv.sin_addr.s_addr = htonl(INADDR_ANY);
	serv.sin_port = port;

	if ((bind(listenfd, (struct sockaddr *) &serv, sizeof(serv))) == -1){
		perror("Bind error");
		exit(1);
	} 
	

	if ((listen(listenfd, 5)) == -1){ // We don't need a huge backlog (arg 2). 
		perror("Send error");
		exit(1);
	}

	printf("\x1B[32mSending on port %d...\033[0m\n", ntohs(serv.sin_port));
	

	for (;;) {
		len = sizeof(cli);
		connfd = accept(listenfd, (struct sockaddr *) &cli, &len);

		inet_ntop(AF_INET, &(cli.sin_addr), ip_buff, INET_ADDRSTRLEN);
		printf("\x1B[32mConnection from %s, port %d\033[0m\n", ip_buff, ntohs(cli.sin_port));

		if ((bytes_sent = send_file(filename_fd, connfd)) < 0)
			printf("\x1B[31mError transferring file.\033[0m\n\e[?25h");
		else
			printf("\x1B[32mFinished transfer to %s\n\%d bytes sent.\033[0m\n\e[?25h", ip_buff, bytes_sent);exit(0);
	}

	free(ip_buff);

	return 0;
}


hermes.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/types.h>
#include "hermes.h"

void print_help(void){
	printf("Hermes - v0.2\n\nhermes <filename> [-h] [-p <port>] \n \
			-h: Prints this help.\n \
			-v: Be verbose.\n \
			-p: Specify a port to listen on. Default 9999.\n");
	exit(0);
}

int main(int argc, char *argv[])
{

	be_verbose = 0;
	FILE *filename_fd;
	char *filename_to_serve;
	int listen_port = htons(9999);

	if (argc < 2)
		print_help();

	char c;
	while ((c = getopt(argc, argv, "hvp:")) != -1) {
		switch(c){

			case 'h':
				print_help();
				break;
			case 'v':
				printf("Being verbose.\n");
				be_verbose = 1;
				break;
			case 'p':
				if ((optarg != NULL) && (optarg[0] != '\0')){

					listen_port = htons(atoi(optarg));

					if (listen_port < 1024 && getuid() > 0){
						printf("Invalid port number. Binding to a port lower than 1024 requires superuser priveleges. You, unfortunately, are not so priveleged. Sorry!\n");
						exit(0);
					}
					else {
						listen_port = htons(atoi(optarg));
						break;
					}
				}
				else break;
			default:
				print_help();
				break;
		}
	}

	// Get the filename, which is the argument without an option.
	if ((argc - optind) == 1)
	{
		if (strlen(argv[argc-1]) > 255)
		{
			printf("Filename too long.\n");
			exit(0);
		}

		if((filename_to_serve = malloc(strlen(argv[argc-1]) + 1)) == NULL)
		{
			perror("Malloc");
			exit(1);
		}
		strncpy(filename_to_serve, argv[argc-1], strlen(argv[argc-1]));
	}


	if ((filename_fd = fopen(filename_to_serve, "r")) != NULL){
		listen_for_inbound_requests(listen_port, filename_fd);
		printf("%d\n", listen_port);
	}

	else
		perror("Error opening file");

	free(filename_to_serve);


	return 0;
}


